lib/rules/jsx-no-constructed-context-values.js

```diff
@@ -0,0 +1,215 @@
+/**
+ * @fileoverview Prevents jsx context provider values from taking values that
+ *               will cause needless rerenders.
+ * @author Dylan Oshima
+ */
+
+'use strict';
+
+const docsUrl = require('../util/docsUrl');
+
+// ------------------------------------------------------------------------------
+// Helpers
+// ------------------------------------------------------------------------------
+
+// Recursively checks if an element is a construction.
+// A construction is a variable that changes reference every render.
+function isConstruction(node, callScope) {
+  switch (node.type) {
+    case 'Literal':
+      if (node.value instanceof RegExp) {
```

Comment:
instanceof is highly unreliable and should be avoided on principle; let's use https://www.npmjs.com/package/is-regex if this is needed.

Reply:
Before we go down that route another method I was using was just checking that there was a [regex property in the literal node](https://astexplorer.net/#/gist/ef24fe782783aedcdf565b24c2585f34/latest) but I had thought the `instanceof` might be safer.

Reply:
I'm fine relying on the AST structure; instanceof isn't ever safe, unfortunately.

Reply:
Could you please elaborate on why it's not safe in this context?

I would have thought that based on the context of a lint rule, we can make some assumptions that imply enough safety:
- The AST is provided by a parser (via ESLint), so it's not some user generated object passed to a function.
- The AST is assumed to conform [to the ESTree spec for literals](https://github.com/estree/estree/blob/master/es5.md#literal)
- The `value` prop is assumed to be one of a set of known of types (string, boolean, number, bigint, null, or RegExp)

Unless I'm wrong, the only case that it's not safe is if someone has purposely crafted a parser for eslint which provides an AST that is not safe?

Reply:
It wouldn't work if a parser created the regex in another realm, like in the `vm` module, but you're right that in this specific use case it's probably fine. I'd still prefer to avoid it to keep up a good coding habit.

Reply:
I didn't know about that API! That makes sense then.